### PR TITLE
[Refactor] Use common sp utils for Qwen3.5 MoE

### PR DESCRIPTION
**Pull request by gcanlin**
## Summary

This PR enables the common sequence-parallel path for Qwen3.5 MoE, keeps sequence-sharded tensors across attention/MoE boundaries, and restores full-token outputs only where required. It also preserves the Qwen3.5 MTP path by normalizing the local shard before gathering the final draft-model output.

## End-to-end flow

```text
Qwen3_5ForConditionalGeneration.forward
  -> Qwen3_5Model.forward
     -> Qwen3NextDecoderLayer.forward
        -> Qwen3NextAttention.forward
           -> sp_shard / sp_reduce_scatter
        -> Qwen3NextSparseMoeBlock.forward
           -> sp_all_gather (only when the MoE input must be restored)
     -> _restore_sequence_parallel_outputs
        -> one packed sp_all_gather for final and auxiliary states

MTP:
Qwen3_5MTP.forward
  -> decoder layer
  -> RMSNorm on the local sequence shard
  -> sp_all_gather
  -> trim sequence padding
```

## Results

Model: `Qwen/Qwen3.5-122B-A10B-FP8`, TP=2, DP=2, EP enabled.

### Non-MTP: main vs. this PR

The throughput numbers below use the steady-state runs (main's final run and the mean of this PR's final three runs).

| Branch | Output throughput | Total token throughput | Mean TTFT | Mean TPOT | GSM8K 5-shot flexible / strict |
|---|---:|---:|---:|---:|---:|
| main | 1,853.15 tok/s | 20,384.67 tok/s | 1,013.50 ms | 13.68 ms | Not separately rerun |
| this PR | 1,843.41 tok/s | 20,277.53 tok/s | 1,117.85 ms | 13.55 ms | 86.66% / 85.14% |

Output throughput differs by **-0.53%**, while mean TPOT improves by **0.94%**.

```bash
vllm bench serve \
  --model Qwen/Qwen3.5-122B-A10B-FP8 \
  --host localhost --port 8000 \
  --dataset-name random \
  --random-input-len 10240 \
  --random-output-len 1024 \
  --num-prompts 100 \
  --max-concurrency 32 \
  --temperature 0 --ignore-eos \
  --num-warmups 4

lm_eval \
  --model local-completions \
  --model_args model=Qwen/Qwen3.5-122B-A10B-FP8,base_url=http://127.0.0.1:8000/v1/completions,tokenizer=Qwen/Qwen3.5-122B-A10B-FP8,num_concurrent=32,max_retries=3,tokenized_requests=False,timeout=600 \
  --tasks gsm8k --num_fewshot 5 --batch_size 1 \
  --gen_kwargs temperature=0,max_gen_toks=256
```

### MTP: this PR

MTP was tested with two speculative tokens. A representative shape-stress run used 127 input tokens, 64 output tokens, 32 prompts, and concurrency 32.

| Branch | Output throughput | Mean TTFT | Mean TPOT | MTP acceptance | GSM8K 5-shot flexible / strict |
|---|---:|---:|---:|---:|---:|
| this PR + MTP | 1,421.63 tok/s | 500.94 ms | 9.20 ms | 78.66% | 86.05% / 84.38% |

All 32 requests completed successfully. Additional input lengths `1`, `3`, and `17` were also tested to cover small and non-divisible sequence shapes.

```bash
vllm serve Qwen/Qwen3.5-122B-A10B-FP8 \
  --tensor-parallel-size 2 \
  --data-parallel-size 2 \
  --enable-expert-parallel \
  --max-model-len 32768 \
  --speculative-config '{"method":"mtp","num_speculative_tokens":2,"max_model_len":32768}'

vllm bench serve \
  --model Qwen/Qwen3.5-122B-A10B-FP8 \
  --host localhost --port 8000 \
  --dataset-name random \
  --random-input-len 127 \
  --random-output-len 64 \
  --num-prompts 32 \
  --max-concurrency 32 \
  --temperature 0 --ignore-eos

lm_eval \
  --model local-completions \
  --model_args model=Qwen/Qwen3.5-122B-A10B-FP8,base_url=http://127.0.0.1:8000/v1/completions,tokenizer=Qwen/Qwen3.5-122B-A10B-FP8,num_concurrent=32,max_retries=3,tokenized_requests=False,timeout=600 \
  --tasks gsm8k --num_fewshot 5 --batch_size 1 \
  --gen_kwargs temperature=0,max_gen_toks=256
```

non-MTP:

(main)
```
local-completions ({'model': 'Qwen/Qwen3.5-122B-A10B-FP8', 'base_url': 'http://127.0.0.1:8000/v1/completions', 'tokenizer': 'Qwen/Qwen3.5-122B-A10B-FP8', 'num_concurrent': 32, 'max_retries': 3, 'tokenized_requests': False, 'timeout': 600}), gen_kwargs: ({'temperature': 0, 'max_gen_toks': 256}), limit: None, num_fewshot: 5, batch_size: 1
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8650|±  |0.0094|
|     |       |strict-match    |     5|exact_match|↑  |0.8423|±  |0.0100|
```

(This PR)

```text
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8666|±  |0.0094|
|     |       |strict-match    |     5|exact_match|↑  |0.8514|±  |0.0098|
```

MTP:

```text
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8605|±  |0.0095|
|     |       |strict-match    |     5|exact_match|↑  |0.8438|±  |0.0100|
```

### CODE REVIEW

File: vllm/models/common/ops/sequence_parallel.py [hunk 1]
```diff
@@ -10,6 +10,39 @@
...
+    return torch.nn.functional.pad(x, pad, value=value)
+
+
+def _sp_pad_compiled_fake(x: torch.Tensor, tp_size: int, value: float) -> torch.Tensor:
+    new_shape = list(x.shape)
+    new_shape[0] += (-new_shape[0]) % tp_size
+    return torch.empty(new_shape, dtype=x.dtype, device=x.device)
+
+
+direct_register_custom_op(
```
Comment (gcanlin) [inline vllm/models/common/ops/sequence_parallel.py:28]: This change is for compile path, which has been removed for Qwen. Can revert this change after we forward to non-compile path.

**Comment by gcanlin** [comment]
cc @ZJY0516

**Review by ZJY0516 — COMMENTED** [review]
We don't want compiled files and non-compiled files share too much code

**Comment by gcanlin** [comment]
Also fix #50681 

```
vllm serve Qwen/Qwen3.6-35B-A3B --tensor-parallel-size 2 --data-parallel-size 2 --enable-expert-parallel --max-model-len 32768 --enforce-eager
```

```
(vllm) root@h200-172:~/vllm-workspace/vllm# curl -sS http://127.0.0.1:8000/v1/chat/completions   --request POST   -H 'Content-Type: application/json'   -d '{
    "model": "Qwen/Qwen3.6-35B-A3B",
    "messages": [
      {
        "role": "user",
        "content": "Explain quantum computing in simple terms:"
      }
    ],
    "temperature": 0,
    "max_tokens": 64,
    "chat_template_kwargs": {
      "enable_thinking": false
    }
  }'
{"id":"chatcmpl-8825268b8175c58e","object":"chat.completion","created":1786553312,"model":"Qwen/Qwen3.6-35B-A3B","choices":[{"index":0,"message":{"role":"assistant","content":"To understand quantum computing, it helps to first look at how our current computers work, and then see how quantum computers break the rules.\n\n### 1. The Classical Computer: Light Switches\nYour smartphone or laptop is a **classical computer**. It processes information using **bits**.\n- A bit is like","refusal":null,"annotations":null,"audio":null,"function_call":null,"reasoning":null},"logprobs":null,"finish_reason":"length","stop_reason":null,"token_ids":null,"routed_experts":null}],"service_tier":null,"system_fingerprint":"vllm-0.26.1rc1.dev631+g5426311d9-tp2-dp2-ep-0f17a4a1","usage":{"prompt_tokens":20,"total_tokens":84,"completion_tokens":64,"prompt_tokens_details":null},"prompt_logprobs":null,"prompt_token_ids":null,"prompt_text":null,"kv_transfer_params":null,"ec_transfer_params":null,"metrics":null}(vllm) root@h200-172:~/vllm-workspace/vllm#
```

**Comment by gcanlin** [comment]
How about this flow:

```
Qwen3NextModel.forward
  │
  ├─ 1. Token embedding
  │    hidden_states: [T, H], replicated across TP ranks
  │
  ├─ 2. sp_padding_mask(...)
  │    Marks the padding tokens added for TP divisibility.
  │    Used when VLLM_MOE_SKIP_PADDING is enabled.
  │
  ├─ 3. sp_shard(hidden_states)
  │    ├─ _sp_pad(...)
  │    │    Pads T to a multiple of the TP size.
  │    └─ Selects the token chunk owned by the current TP rank.
  │
  │    [T_pad, H] → [T_pad / TP, H]
  │
  ├─ 4. Qwen3NextDecoderLayer.forward × N
  │    │
  │    ├─ input_layernorm(sharded states)
  │    │
  │    ├─ sp_all_gather(hidden_states)
  │    │    ├─ Uses custom_all_gather when available
  │    │    └─ Falls back to the regular TP all-gather
  │    │
  │    │    [T_pad / TP, H] → [T_pad, H] → [:T]
  │    │
  │    ├─ Attention over the full token set
  │    │    ├─ QwenGatedDeltaNetAttention, or
  │    │    └─ Qwen3NextAttention
  │    │
  │    │    o_proj.reduce_results=False, so every rank produces
  │    │    its TP-local partial attention result.
  │    │
  │    ├─ sp_reduce_scatter(attention_output)
  │    │    ├─ Pads the token dimension when necessary
  │    │    ├─ Uses custom_reduce_scatter when available
  │    │    └─ Falls back to the regular TP reduce-scatter
  │    │
  │    │    This collective simultaneously:
  │    │      1. sums the TP-local partial results; and
  │    │      2. shards the result along the token dimension.
  │    │
  │    │    [T, H] → [T_pad / TP, H]
  │    │
  │    ├─ post_attention_layernorm(sharded states)
  │    │
  │    ├─ FFN
  │    │    │
  │    │    ├─ MoE layer
  │    │    │    Qwen3NextSparseMoeBlock.forward(
  │    │    │        already_sequence_parallel=True
  │    │    │    )
  │    │    │
  │    │    │    The input is already sequence-sharded, so the MoE
  │    │    │    block does not shard or gather it again.
  │    │    │
  │    │    └─ Dense layer
  │    │         Qwen3NextMLP.forward(sharded states)
  │    │
  │    │         It is constructed with is_sequence_parallel=True:
  │    │           gate_up_proj.disable_tp=True
  │    │           down_proj.disable_tp=True
  │    │
  │    │         Each TP rank holds a complete copy of the dense MLP
  │    │         weights and processes its local token shard directly.
  │    │
  │    └─ Return sharded hidden_states and residual
  │
  ├─ 5. Final norm over sharded states
  │
  └─ 6. sp_restore_outputs(...)
       ├─ Packs the final and auxiliary hidden states along the hidden dimension
       ├─ Restores them with a single sp_all_gather
       ├─ Removes SP padding with [:full_num_tokens]
       └─ Splits the packed tensor back into individual outputs

       [T_pad / TP, H] → [T, H]
```


MTP:

```
 Qwen MTP forward
  │
  ├─ Normalize inputs_embeds and previous_hidden_states
  ├─ Concatenate and project with fc
  ├─ sp_shard(hidden_states)
  ├─ MTP decoder layer
  │    ├─ sp_all_gather before attention
  │    ├─ Attention over the full token set
  │    ├─ sp_reduce_scatter after attention
  │    └─ MoE FFN over sharded states
  ├─ Final norm over sharded states
  └─ sp_all_gather → remove padding → full MTP output
```
